File: packages/contracts/v1/market_status.go
package v1

// MarketStatusEvent represents a market status change event sent via WebSocket.
// This is used to notify clients when markets open or close.
type MarketStatusEvent struct {
	Type       string `json:"type"`        // Always "market_status"
	AssetClass string `json:"asset_class"` // "forex", "crypto", "stocks", "commodities", "mixed"
	Status     string `json:"status"`      // "open" or "closed"
	Reason     string `json:"reason,omitempty"`
	ReopensAt  string `json:"reopens_at,omitempty"` // RFC3339 format when market will reopen
	ClosesAt   string `json:"closes_at,omitempty"`  // RFC3339 format when market will close
	Ts         int64  `json:"ts"`
}

// MarketStatus represents the current status of a market.
type MarketStatus struct {
	AssetClass string  `json:"asset_class"`
	IsOpen     bool    `json:"is_open"`
	Reason     string  `json:"reason,omitempty"`
	NextOpen   *string `json:"next_open,omitempty"`  // RFC3339 format
	NextClose  *string `json:"next_close,omitempty"` // RFC3339 format
	Override   *string `json:"override,omitempty"`   // If there's an active override
}

// MarketOverride represents a manual override for special events.
type MarketOverride struct {
	AssetClass string `json:"asset_class"`
	Status     string `json:"status"` // "open" or "closed"
	Reason     string `json:"reason"`
	ExpiresAt  string `json:"expires_at"` // RFC3339 format
	CreatedBy  string `json:"created_by"`
	CreatedAt  string `json:"created_at"` // RFC3339 format
}

// MarketHoursConfig represents the market hours configuration for an asset class.
type MarketHoursConfig struct {
	AssetClass string            `json:"asset_class"`
	OpenTime   *MarketTimeSpec   `json:"open_time,omitempty"`  // nil if always open
	CloseTime  *MarketTimeSpec   `json:"close_time,omitempty"` // nil if never closes
	AlwaysOpen bool              `json:"always_open"`
	Holidays   []string          `json:"holidays"` // List of holiday dates in YYYY-MM-DD format
}

// MarketTimeSpec represents a specific day and time for market open/close.
type MarketTimeSpec struct {
	Day      string `json:"day"`      // Day of week (e.g., "Sunday", "Friday")
	Time     string `json:"time"`     // Time in HH:MM format
	Timezone string `json:"timezone"` // IANA timezone (e.g., "UTC", "America/New_York")
}

// MarketStatusRequest is used by admin to query market status.
type MarketStatusRequest struct {
	AssetClass string `json:"asset_class,omitempty"` // If empty, returns all asset classes
}

// MarketStatusResponse contains the market status for one or more asset classes.
type MarketStatusResponse struct {
	Statuses []MarketStatus `json:"statuses"`
}

// SetOverrideRequest is used by admin to set a manual market override.
type SetOverrideRequest struct {
	AssetClass string `json:"asset_class"`
	Status     string `json:"status"`     // "open" or "closed"
	Reason     string `json:"reason"`
	ExpiresAt  string `json:"expires_at"` // RFC3339 format
}

// ValidateContestTimesRequest is used to validate contest times against market hours.
type ValidateContestTimesRequest struct {
	AssetClass string `json:"asset_class"`
	StartsAt   string `json:"starts_at"` // RFC3339 format
	EndsAt     string `json:"ends_at"`   // RFC3339 format
}

// ValidateContestTimesResponse contains validation results.
type ValidateContestTimesResponse struct {
	Valid  bool   `json:"valid"`
	Reason string `json:"reason,omitempty"` // Why validation failed
}

File: packages/contracts/ts/v1/market-status.ts
/**
 * MarketStatusEvent represents a market status change event sent via WebSocket.
 * This is used to notify clients when markets open or close.
 */
export interface MarketStatusEvent {
  type: 'market_status';
  asset_class: string;  // "forex", "crypto", "stocks", "commodities", "mixed"
  status: 'open' | 'closed';
  reason?: string;
  reopens_at?: string;  // RFC3339 format when market will reopen
  closes_at?: string;   // RFC3339 format when market will close
  ts: number;
}

/**
 * MarketStatus represents the current status of a market.
 */
export interface MarketStatus {
  asset_class: string;
  is_open: boolean;
  reason?: string;
  next_open?: string;   // RFC3339 format
  next_close?: string;  // RFC3339 format
  override?: string;    // If there's an active override
}

/**
 * MarketOverride represents a manual override for special events.
 */
export interface MarketOverride {
  asset_class: string;
  status: 'open' | 'closed';
  reason: string;
  expires_at: string;   // RFC3339 format
  created_by: string;
  created_at: string;   // RFC3339 format
}

/**
 * MarketTimeSpec represents a specific day and time for market open/close.
 */
export interface MarketTimeSpec {
  day: string;      // Day of week (e.g., "Sunday", "Friday")
  time: string;     // Time in HH:MM format
  timezone: string; // IANA timezone (e.g., "UTC", "America/New_York")
}

/**
 * MarketHoursConfig represents the market hours configuration for an asset class.
 */
export interface MarketHoursConfig {
  asset_class: string;
  open_time?: MarketTimeSpec;   // undefined if always open
  close_time?: MarketTimeSpec;  // undefined if never closes
  always_open: boolean;
  holidays: string[];           // List of holiday dates in YYYY-MM-DD format
}

/**
 * MarketStatusRequest is used by admin to query market status.
 */
export interface MarketStatusRequest {
  asset_class?: string; // If empty, returns all asset classes
}

/**
 * MarketStatusResponse contains the market status for one or more asset classes.
 */
export interface MarketStatusResponse {
  statuses: MarketStatus[];
}

/**
 * SetOverrideRequest is used by admin to set a manual market override.
 */
export interface SetOverrideRequest {
  asset_class: string;
  status: 'open' | 'closed';
  reason: string;
  expires_at: string; // RFC3339 format
}

/**
 * ValidateContestTimesRequest is used to validate contest times against market hours.
 */
export interface ValidateContestTimesRequest {
  asset_class: string;
  starts_at: string; // RFC3339 format
  ends_at: string;   // RFC3339 format
}

/**
 * ValidateContestTimesResponse contains validation results.
 */
export interface ValidateContestTimesResponse {
  valid: boolean;
  reason?: string; // Why validation failed
}

/**
 * Helper to check if a market status event indicates the market is closed.
 */
export function isMarketClosed(event: MarketStatusEvent): boolean {
  return event.status === 'closed';
}

/**
 * Helper to format market reopen time as a human-readable string.
 */
export function formatReopensAt(event: MarketStatusEvent): string | null {
  if (!event.reopens_at) return null;
  try {
    const date = new Date(event.reopens_at);
    return date.toLocaleString();
  } catch {
    return event.reopens_at;
  }
}
